docs: Add separate information about scripts (#4643)
scalameta/metals: docs/features/scripts.md

---
id: scripts
title: Scripts support
---

### Working with a script

Whenever, Metals opens a script file with `*.sc` extension, but only when it's

> not `*.worksheet.sc` or a build server is not already started for it, users will

Reviewer: ```suggestion
not `*.worksheet.sc` or a build server has not already started for it, users will
```

be prompted to choose whether they want to use [Ammonite](http://ammonite.io/)

> or [ScalaCli](https://scala-cli.virtuslab.org/) for using within the it.

Reviewer: ```suggestion
or [ScalaCli](https://scala-cli.virtuslab.org/) for usage within it.
```

Reviewer 2: I actually think this is worded a bit oddly. I'l recommend actually changing the whole ending to something like:

> for it.

If you want to keep it really simple, but if you want to be more detailed something like:

> the power the script.

> or [ScalaCli](https://scala-cli.virtuslab.org/) for using within the it.

Reviewer 2: Just a nit, but I never see `ScalaCLI` all as one word like this. Would it be more accurate to say `Scala CLI`?

Author: Actually it's `Scala CLI` based on what we wrote on the website :sweat_smile:

Author: Changed everywhere I think!

![scala-cli](https://i.imgur.com/ghR1Src.gif)

Scripts are usually used for smaller programs and the main difference is that
statements can be put on top level without a need for additional objects or
methods.

> Once user have chosen a specific scripting tool, Metals will also suggest to

Reviewer 2: ```suggestion
Once a user has chosen a specific scripting tool, Metals will also suggest to
```

Author: done!

import all newly opened scripts automatically or to always to do that manually.

Both tools will start a build server in the background that will be able to
compile your code and provide Metals with all the necessary information.

> ScalaCLI scripts can also be used standalone when ScalaCli generated `.bsp`

Reviewer 2: I see both `ScalaCLI` and `ScalaCli`. Whatever you choose you should use the same throughout.

---
id: scripts
title: Scripts support
---

### Working with a script

Whenever, Metals opens a script file with `*.sc` extension, but only when it's
not `*.worksheet.sc` or a build server is not already started for it, users will
be prompted to choose whether they want to use [Ammonite](http://ammonite.io/)
or [Scala CLI](https://scala-cli.virtuslab.org/) to power the script.

![scala-cli](https://i.imgur.com/ghR1Src.gif)

Scripts are usually used for smaller programs and the main difference is that
statements can be put on top level without a need for additional objects or
methods.

Once a user has chosen a specific scripting tool, Metals will also suggest to
import all newly opened scripts automatically or to always to do that manually.

Both tools will start a build server in the background that will be able to
compile your code and provide Metals with all the necessary information.

> ScalaCLI scripts can also be used standalone when Scala CLI generated `.bsp`

Reviewer 2: ```suggestion
Scala CLI scripts can also be used standalone when Scala CLI generated `.bsp`
```

script.

### Advanced information

Additionally, advanced users can start the underlying build server manually with
`Metals: Start Scala CLI BSP server` (`scala-cli-start`) or
`Metals: Start Ammonite BSP server` (`ammonite-start`). They will also be able
to stop it with `Metals: Stop Scala CLI BSP server`(`scala-cli-stop`) or
`Metals: Stop Ammonite BSP server`(ammonite-stop). These commands can be used
for more fine grained control when to turn on or of scripting support. This is
especially useful since the additional build server running underneath can take
up some additional resources.

If the script is in a dedicated folder, by default we will treat all the scripts
and scala files in that directory as ones that can be used together. So you
would be able to import method and classes from those files. However, if the
script is contained within a directory that also contains other sources, that
script will be treated as a standalone one in order to avoid flaky compilation
errors coming from normal files in the workspace.

Current limitations can be found:

- [here for Ammonite](https://github.com/scalameta/metals/issues?q=is%3Aopen+is%3Aissue+label%3A%22ammonite+support%22)

> - [here for ScalaCLI](https://github.com/scalameta/metals/issues?q=is%3Aopen+is%3Aissue+label%3Ascala-cli)

Reviewer 2: ```suggestion
- [here for Scala CLI](https://github.com/scalameta/metals/issues?q=is%3Aopen+is%3Aissue+label%3Ascala-cli)
```
